README.md
# Rpi-Repo

## history:
echo 'HISTTIMEFORMAT="%d/%m/%y %T "' >> ~/.bashrc \
source ~/.bashrc

## ssh:
mkdir .ssh \
ssh-keygen -t rsa -C "oliwier.pszeniczko@gmail.com" -f "id_rsa_ollieux_rpi" \
eval "$(ssh-agent -s)" \
ssh-add ~/.ssh/id_rsa_ollieux_rpi \
ssh -T git@github.com

## catalog -> git:
git init\
git remote add origin <> \
git config user.email oliwier.pszeniczko@gmail.com \
git config user.name Ollieux \
git pull origin main

## git -> catalog:
git clone http:<> \
git config user.email oliwier.pszeniczko@gmail.com \
git config user.name Ollieux \

##Open-cv:
sudo apt-get install python3-pip python3-virtualenv

mkdir project \
cd project/
virtualenv env

sudo apt install -y build-essential cmake pkg-config libjpeg-dev libtiff5-dev libpng-dev libavcodec-dev libavformat-dev libswscale-dev libv4l-dev libxvidcore-dev libx264-dev libfontconfig1-dev libcairo2-dev libgdk-pixbuf2.0-dev libpango1.0-dev libgtk2.0-dev libgtk-3-dev libatlas-base-dev gfortran libhdf5-dev libhdf5-serial-dev libhdf5-103 libqt5gui5 libqt5webkit5 libqt5test5 python3-pyqt5 python3-dev

pip install --upgrade pip setuptools wheel

pip install opencv-python==4.5.3.56

## Owner change:
git rebase -i --root\
git commit --amend --author="Ollieux <oliwier.pszeniczko@gmail.com>" --no-edit\
git rebase --continue\

## Wifi
sudo apt-get install -y linux-headers \
git clone https://github.com/lwfinger/rtl8188eu.git \
cd rtl8188eu/ \
make \
sudo make install \
sudo reboot \
sudo nano /etc/network/interfaces \
sudo nano /etc/wpa_supplicant.conf

## additonal
sudo apt-get install conky \
pip install imutils

python -u Rpi-Repo/arduino.py | tee calibration.txt


history.txt
    3  06/02/23 22:03:24 sudo motion
    4  06/02/23 22:04:32 sudo motion stop
    5  06/02/23 22:08:49 sudo reboot
    6  06/02/23 22:19:08 sudo service motion restart
    7  06/02/23 22:19:14 sudo service motion start
    8  06/02/23 22:20:33 sudo nano /etc/default/motion
    9  06/02/23 22:21:06 sudo service motion restart
   10  06/02/23 22:21:09 sudo service motion start
   11  06/02/23 22:30:23 sudo systemctl start motion
   12  06/02/23 22:31:11 sudo systemctl stop motion
   13  06/02/23 22:31:24 sudo reboot
   14  06/02/23 22:11:29 conky
   15  06/02/23 22:34:55 sudo systemctl start motion
   16  06/02/23 22:35:08 systemctl start motion
   17  06/02/23 22:35:35 sudo systemctl stop motion
   18  06/02/23 22:35:48 sudo systemctl restart motion
   19  06/02/23 22:35:55 motion start
   20  06/02/23 22:36:04 sudo motion stop
   21  06/02/23 22:36:58 sudo nano /etc/motion/motion.conf 
   22  06/02/23 22:54:17 sudo systemctl restart motion
   23  06/02/23 22:54:39 sudo systemctl stop  motion
   24  06/02/23 22:54:52 ls /etc/default/
   25  06/02/23 22:55:19 cat /etc/default/ssh 
   26  06/02/23 22:55:40 cat /etc/default/motion 
   27  06/02/23 22:56:10 sudo reboot
   28  06/02/23 23:00:01 sudo killall conky
   29  06/02/23 23:00:14 sudo systemctl start motion
   30  06/02/23 23:00:27 sudo motion 
   31  06/02/23 23:03:03 sudo systemctl stop motion.service
   32  06/02/23 23:05:02 sudo systemctl restart motion.service
   33  06/02/23 23:05:15 sudo reboot
   34  06/02/23 22:58:47 conky
   35  06/02/23 23:13:58 sudo rm /var/lib/motion/*
   36  06/02/23 23:14:31 sudo systemctl start motion.service 
   37  06/02/23 23:14:58 sudo motion
   38  06/02/23 23:16:26 sudo nano /etc/motion/motion.conf 
   39  06/02/23 23:20:33 sudo reboot
   40  06/02/23 23:23:57 sudo systemctl start motion.service
   41  06/02/23 23:24:07 sudo motion
   42  07/02/23 00:08:10 sudo reboot
   43  06/02/23 23:23:22 conky
   44  06/02/23 23:31:33 history
   45  06/02/23 23:37:11 ./git.sh 
   46  06/02/23 23:37:25 source project/env/bin/activate
   47  06/02/23 23:37:38 python Rpi-Repo/Webcam/webcam_fps.py 
   48  06/02/23 23:44:16 ./git.sh 
   49  06/02/23 23:47:41 python Rpi-Repo/Webcam/webcam_fps.py 
   50  07/02/23 13:12:55 conky
   51  07/02/23 12:56:20 sudo motion
   52  07/02/23 16:07:58 sudo nano /etc/motion/motion.conf 
   53  07/02/23 19:28:25 source project/env/bin/activate
   54  07/02/23 19:28:39 python Rpi-Repo/Webcam/webcam_fps.py 
   55  07/02/23 19:49:51 deactivate 
   56  07/02/23 19:49:55 sudo reboot
   57  07/02/23 19:53:52 conky
   58  08/02/23 00:50:02 sudo reboot
   59  07/02/23 19:54:13 sudo motion
   60  08/02/23 00:53:10 conky
   61  08/02/23 01:02:44 sudo reboot
   62  08/02/23 14:00:23 conky
   63  08/02/23 14:03:16 source project/env/bin/activate
   64  08/02/23 14:15:01 python Rpi-Repo/motorApp75.py 
   65  08/02/23 14:16:24 deactivate 
   66  08/02/23 14:17:59 python Rpi-Repo/motorApp75.py 
   67  08/02/23 14:20:58 history
   68  08/02/23 14:27:48 python Rpi-Repo/motorApp75.py 
   69  09/02/23 01:09:01 sudo shutdown -h now
   70  09/02/23 10:10:01 conky
   71  09/02/23 10:11:01 sudo motion
   72  09/02/23 10:11:36 source project/env/bin/activate
   73  09/02/23 10:11:53 python Rpi-Repo/Webcam/webcam_fps.py 
   74  09/02/23 10:15:08 sudo reboot
   75  09/02/23 10:13:34 conky
   76  09/02/23 11:23:38 ./git
   77  09/02/23 11:23:45 ./git.sh
   78  09/02/23 10:17:55 source project/env/bin/activate
   79  09/02/23 10:18:13 python Rpi-Repo/Webcam/webcam_fps.py 
   80  09/02/23 11:41:55 python Rpi-Repo/Webcam/webcam_fps.py 
   81  09/02/23 11:45:57 cat Rpi-Repo/Fire/fire_detection_class.py
   82  09/02/23 11:47:42 python Rpi-Repo/Fire/fire_detection_class.py
   83  09/02/23 11:57:41 python Rpi-Repo/Webcam/webcam_fps.py 
   84  09/02/23 11:59:53 ./git.sh 
   85  09/02/23 12:00:01 python Rpi-Repo/Fire/fire_detection_class.py
   86  09/02/23 12:01:16 ./git.sh 
   87  09/02/23 12:01:22 python Rpi-Repo/Fire/fire_detection_class.py
   88  09/02/23 12:04:39 ./git.sh 
   89  09/02/23 12:04:52 python Rpi-Repo/Fire/fire_detection_class.py
   90  09/02/23 12:08:26 ./git.sh 
   91  09/02/23 12:08:35 cat Rpi-Repo/Fire/fire_detection_class.py
   92  09/02/23 12:08:50 python Rpi-Repo/Fire/fire_detection_class.py
   93  09/02/23 12:11:32 ./git.sh 
   94  09/02/23 12:11:38 cat Rpi-Repo/Fire/fire_detection_class.py
   95  09/02/23 12:12:07 cd ~/Rpi-Repo/Fire/fire_detection.xml
   96  09/02/23 12:12:16 cd ~/Rpi-Repo/Fire/
   97  09/02/23 12:12:28 cd ~
   98  09/02/23 12:13:52 python Rpi-Repo/Fire/fire_detection_class.py
   99  09/02/23 12:20:49 ./git.sh 
  100  09/02/23 12:21:03 cat Rpi-Repo/Fire/fire_detection_class.py
  101  09/02/23 12:21:23 python Rpi-Repo/Fire/fire_detection_class.py
  102  09/02/23 12:23:24 ./git.sh
  103  09/02/23 12:23:38 python Rpi-Repo/Fire/fire_detection_class.py
  104  09/02/23 12:24:53 python
  105  09/02/23 12:42:43 ./git.sh
  106  09/02/23 12:43:03 python Rpi-Repo/Fire/fire_detection_class.py 
  107  09/02/23 12:44:02 python Rpi-Repo/Webcam/webcam_fps.py 
  108  09/02/23 12:54:26 ./git.sh
  109  09/02/23 12:55:00 cat Rpi-Repo/Fire/fire_detection_class.py 
  110  09/02/23 13:02:36 deactivate 
  111  09/02/23 12:24:40 python
  112  09/02/23 12:26:46 ls /home/rpi/project/env/lib/python3.9/site-packages/cv2/data/
  113  09/02/23 12:52:13 history
  114  09/02/23 13:03:11 ./git.sh 
  115  09/02/23 13:03:27 source project/env/bin/activate
  116  09/02/23 13:04:03  
  117  09/02/23 13:05:11 cat Rpi-Repo/Fire/fire_detection.xml
  118  09/02/23 13:14:20 python Rpi-Repo/Fire/fire_detection_class.py 
  119  09/02/23 13:13:44 ./git.sh 
  120  09/02/23 13:20:44 source project/env/bin/activate
  121  09/02/23 13:20:49 python Rpi-Repo/Fire/fire_detection_class.py 
  122  09/02/23 13:22:41 deactivate 
  123  09/02/23 13:22:49 ./git.sh 
  124  09/02/23 13:22:54 source project/env/bin/activate
  125  09/02/23 13:22:57 python Rpi-Repo/Fire/fire_detection_class.py 
  126  09/02/23 13:34:19 deactivate 

App/main.py
import os
import socket
import struct
import threading
import time
import serial
import cv2
import queue
import firebase_admin
import imutils
import RPi.GPIO as GPIO
import w1thermsensor
from firebase_admin import credentials, messaging


def send_notification(factor, _):

    title = "Warning!"
    temperature = sensor.get_temperature()

    if factor == "fire":
        message = "fire detected, temperature: " + str(temperature) + "*C"

    elif factor == "smoke":

        message = "smoke detected, temperature: " + str(temperature) + "*C"

    else:
        message = ""

    msg = messaging.Message(
        notification=messaging.Notification(
            title=title,
            body=message
        ),
        token=registration_token,
    )

    messaging.send(msg)


def send_controls():

    arduino = serial.Serial(port='/dev/ttyUSB0', baudrate=115200)

    while True:
        data = data_queue.get()
        # data += '#'
        arduino.write(data.encode())


def receive_data():

    while True:

        try:
            data = connection.recv(SIZE).decode("utf-8")
            if not data:

                print("null")
                break

            data_queue.put(data)

        except Exception as e:
            break


def send_frame():

    global gtemperature
    temp = gtemperature

    threading.Thread(target=read_temp, ).start()

    while True:

        sframe = send_frame_queue.get()

        try:
            temp = temp_queue.get(block=False)

        except queue.Empty:
            pass

        finally:
            gtemperature = temp

        sframe = cv2.putText(sframe, 'temp: ' + str(temp) + '*C', (10, 10), cv2.FONT_HERSHEY_SIMPLEX, 0.4, (255, 0, 255), 1)

        try:
            data = cv2.imencode('.jpg', sframe)[1].tobytes()
            connection.sendall(struct.pack("!i", len(data)) + data)
        except Exception as e:
            print(e)
            break


def capture_frame():

    global capturing

    cap = cv2.VideoCapture(0)
    cap.set(3, WIDTH)
    cap.set(4, HEIGHT)
    cap.set(5, FPS)
    capturing = True
    while cap.isOpened():

        ret, frame = cap.read()
        frame = cv2.rotate(frame, cv2.ROTATE_90_COUNTERCLOCKWISE)

        if detecting:

            detect_frame_queue.put(frame)

        if connected:

            send_frame_queue.put(frame)

        # #TODO: final out
        cv2.imshow("Capture", frame)
        key = cv2.waitKey(1) & 0xFF
        if key == ord('q'):

            break

        if not running:

            break

    capturing = False
    cap.release()


def detect_smoke():

    while True:

        if not GPIO.input(18):

            global smoke_notified, smoke_time
            if smoke_notified:

                if time.time() - smoke_time > NOTIFICATION_TIMER:

                    smoke_notified = False

            if not smoke_notified:

                threading.Thread(target=send_notification, args=("smoke", "")).start()
                smoke_time = time.time()
                smoke_notified = True

        time.sleep(1)


def detect_fire():

    global detecting
    detecting = True
    while running:

        dframe = detect_frame_queue.get()

        gray = cv2.cvtColor(dframe, cv2.COLOR_BGR2GRAY)
        flames = fire_cascade.detectMultiScale(gray, scaleFactor=1.3, minNeighbors=5)
        for (x, y, w, h) in flames:

            # # TODO: final out
            cv2.rectangle(dframe, (x, y), (x + w, y + h), (255, 0, 0), 2)
            # print("fire detected")

            global fire_notified, fire_time

            if fire_notified:

                if time.time() - fire_time > NOTIFICATION_TIMER:

                    fire_notified = False

            if not fire_notified:

                threading.Thread(target=send_notification, args=("fire", "")).start()
                fire_time = time.time()
                fire_notified = True

        # # TODO: final out
        cv2.imshow("Detect", dframe)
        key = cv2.waitKey(1) & 0xFF
        if key == ord('q'):
            detecting = False
            break


def read_temp():

    while connected:
        read = sensor.get_temperature()
        temp_queue.put(read)
        time.sleep(1)


def button_callback(channel):

    global running
    running = False
    GPIO.cleanup()
    os.system("sudo shutdown -h now")


GPIO.setwarnings(False)

GPIO.setmode(GPIO.BCM)
GPIO.setup(18, GPIO.IN)
GPIO.setup(17, GPIO.IN, pull_up_down=GPIO.PUD_DOWN)
GPIO.add_event_detect(17, GPIO.RISING, callback=button_callback)

sensor = w1thermsensor.W1ThermSensor()

host = socket.gethostbyname(socket.gethostname() + ".local")

port = 9977
server_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
server_socket.bind((host, port))
server_socket.listen(1)

fire_cascade = cv2.CascadeClassifier(cv2.data.haarcascades + 'fire_detection.xml')

cred = credentials.Certificate('/home/rpi/Rpi-Repo/Notification/pushnotifcationtest-f5539-firebase-adminsdk-2ssjk-6ca36caa53.json')
firebase_admin.initialize_app(cred)

registration_token = 'epZx5w_RToGbXxrpEjeMXN:APA91bE3tEshbgB0_T0yMa28AeFkPOCmt4-HxIYwCk4q24mhEqIMvfp1VlEPPfgal0c7JTH3wCTPH-GvxrDgLHuwbmOGvPjAdY4SZ6PXwgCuU6YElAGAaorIg-vBfBh4dr-Dz0FaSrmH'

WIDTH = 320
HEIGHT = 240
FPS = 10
SIZE = 96
NOTIFICATION_TIMER = 30

capturing = False
detecting = False
connected = False
fire_notified = False
smoke_notified = False
fire_time = 0
smoke_time = 0
gtemperature = 0
running = True

send_frame_queue = queue.Queue()
detect_frame_queue = queue.Queue()
data_queue = queue.Queue()
temp_queue = queue.Queue()


capture_thread = threading.Thread(target=capture_frame, )
capture_thread.start()


while not capturing:
    
    pass


fire_thread = threading.Thread(target=detect_fire, )
fire_thread.start()

controls_thread = threading.Thread(target=send_controls, )
controls_thread.start()

smoke_thread = threading.Thread(target=detect_smoke, )
smoke_thread.start()


while True:
    try:
        connection, client_address = server_socket.accept()
        connected = True
        receive_thread = threading.Thread(target=receive_data, )
        receive_thread.start()
        send_thread = threading.Thread(target=send_frame, )
        send_thread.start()
        receive_thread.join()
        send_thread.join()
        connected = False

    except KeyboardInterrupt as e:
        print(e)
        running = False
        GPIO.cleanup()
        break

#TODO: pre final
# git checkout main
# git checkout -b pc
# git merge main albo nie
# git checkout main
# git merge dev-rpi
# TODOs final/final out
